spicy/toolchain/src/ast/types/unit-items/field.cc

```diff
@@ -59,7 +59,45 @@ Type spicy::type::unit::item::Field::itemType() const {
                 return hilti::type::Vector(x->elementType(), x->meta());
 
             return t;
-        });
+        };
+
+        if ( convert->second ) {
+            // For a unit-level &convert, we derive the type of the expression
+            // from the __convert() function that the code generator creates to
+            // evaluate it. That function has an "auto" return type that will
+            // eventually be resolved to the correct type.
+            //
+            // This approach is bit messy, but ensures that "self" inside the
+            // expression gets resolved correctly independent of where the
+            // itemType() that this method returns will be used. (Which
+            // otherwise could be a problem because of our current way of
+            // dynamically computing types. This will become more
+            // straight-forward once we clean up type resolving.)
+            auto& unit = convert->second->as<type::Unit>();
+            auto t = hilti::builder::typeByID(ID(*unit.typeID()));
+            return type::Computed(t, [adapt_type](Node& n) -> Type {
+                auto t = n.as<Type>();
+
+                if ( auto x = t.tryAs<hilti::type::ValueReference>() )
+                    t = x->dereferencedType();
+
+                auto st = t.tryAs<hilti::type::Struct>();
+                if ( ! st )
+                    return type::unknown;
+
+                auto field = st->field("__convert");
+                assert(field);
+
+                auto f = field->type().as<type::Function>();
+                return adapt_type(f.result().type());
```

Comment:
Could you add a comment explaining why we go via the `__convert` field for unit-level converts, but directly via the expression for field-level converts? I naively would have expected these to take a similar route with the unit-level path possibly taking the unit type into account during resolution or similar.

Reply:
There is no __convert for field-level attributes, we just added that to get around typing issues for unit-level. I'll add a note.